library/core/src/main/java/com/google/android/exoplayer2/ExoPlayerImpl.java

```diff
@@ -80,12 +81,28 @@
    */
   @SuppressLint("HandlerLeak")
   public ExoPlayerImpl(
-      Renderer[] renderers, TrackSelector trackSelector, LoadControl loadControl, Clock clock) {
+          Renderer[] renderers, TrackSelector trackSelector, LoadControl loadControl, Clock clock) {
+    this(renderers, trackSelector, loadControl, clock, null);
+  }
+
+  /**
+   * Constructs an instance. Must be called from a thread that has an associated {@link Looper}.
+   *
+   * @param renderers The {@link Renderer}s that will be used by the instance.
+   * @param trackSelector The {@link TrackSelector} that will be used by the instance.
+   * @param loadControl The {@link LoadControl} that will be used by the instance.
+   * @param clock The {@link Clock} that will be used by the instance.
+   */
+  @SuppressLint("HandlerLeak")
+  public ExoPlayerImpl(
+      Renderer[] renderers, TrackSelector trackSelector, LoadControl loadControl, Clock clock,
+      VideoComponent videoComponent) {
```

Comment:
I agree, a log isn't enough in this case. In addition to what you have we should probably bubble up to the ExoPlayer that the uri failed to play, otherwise we can get in a state where we are waiting for a player state to change that never does (maybe the Assert throwing an exception is enough?)

Reply:
I'm not sure if asserting exceptions are the best solution. I have been looking at how this problem is solved in other media source (e.g. HlsMediaSource) and it use the EventDispatcher to notify if an error occurs during the loading of the manifest.

What do you think about setting it as the following example:

    @Override
    public void prepareSourceInternal(ExoPlayer player, boolean isTopLevelSource) {
        Player.VideoComponent videoComponent = player.getVideoComponent();
        Assertions.checkNotNull(videoComponent, "VideoComponent is null.");

        client = new Client.Builder(factory).uri(uri).listener(this).build();

        try {

            client.open();
            videoComponent.addVideoListener(client.session());

        } catch (IOException e) {
            eventDispatcher.loadError(new DataSpec(uri), C.DATA_TYPE_MEDIA_INITIALIZATION, 0, 0,
                    0, new IOException("Client open failed.", e.getCause()), false);
        }
    }

Reply:
Sorry, I forgot the creation of EventDispatcher object:

    @Override
    public void prepareSourceInternal(ExoPlayer player, boolean isTopLevelSource) {
        Player.VideoComponent videoComponent = player.getVideoComponent();
        Assertions.checkNotNull(videoComponent, "VideoComponent is null");

        client = new Client.Builder(factory).uri(uri).listener(this).build();
        EventDispatcher eventDispatcher = createEventDispatcher(/* mediaPeriodId= */ null);

        try {

            client.open();
            videoComponent.addVideoListener(client.session());

        } catch (IOException e) {
            eventDispatcher.loadError(new DataSpec(uri), C.DATA_TYPE_MEDIA_INITIALIZATION, 0, 0,
                    0, new IOException("Client open failed.", e.getCause()), false);
        }
    }


The output console is the following:

I/ExoPlayerImpl: Init 2cd70d80 [ExoPlayerLib/2.7.3] [HWEC6108V9, EC6108V9-01, HUAWEI, 19]
D/EventLogger: state [0.00, true, I]
D/EventLogger: state [0.01, true, B]
E/ViewRootImpl: 201409092 enable acc for view=com.google.android.exoplayer2.demo/com.google.android.exoplayer2.demo.PlayerActivity
E/EventLogger: internalError [3.02, loadError]
               java.io.IOException: Client open failed.
                   at com.google.android.exoplayer2.source.rtsp.RtspMediaSource.prepareSourceInternal(RtspMediaSource.java:196)
                   at com.google.android.exoplayer2.source.BaseMediaSource.prepareSource(BaseMediaSource.java:137)
                   at com.google.android.exoplayer2.ExoPlayerImplInternal.prepareInternal(ExoPlayerImplInternal.java:396)
                   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:286)
                   at android.os.Handler.dispatchMessage(Handler.java:98)
                   at android.os.Looper.loop(Looper.java:136)
                   at android.os.HandlerThread.run(HandlerThread.java:61)
                Caused by: libcore.io.ErrnoException: isConnected failed: EHOSTUNREACH (No route to host)
                   at libcore.io.IoBridge.isConnected(IoBridge.java:208)
                   at libcore.io.IoBridge.connectErrno(IoBridge.java:161)
                   at libcore.io.IoBridge.connect(IoBridge.java:112)
                   at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:192)
                   at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:459)
                   at java.net.Socket.connect(Socket.java:843)
                   at com.google.android.exoplayer2.source.rtsp.api.Dispatcher.connect(Dispatcher.java:86)
                   at com.google.android.exoplayer2.source.rtsp.api.Client.open(Client.java:146)
                   at com.google.android.exoplayer2.source.rtsp.RtspMediaSource.prepareSourceInternal(RtspMediaSource.java:191)
                   at com.google.android.exoplayer2.source.BaseMediaSource.prepareSource(BaseMediaSource.java:137) 
                   at com.google.android.exoplayer2.ExoPlayerImplInternal.prepareInternal(ExoPlayerImplInternal.java:396) 
                   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:286) 
                   at android.os.Handler.dispatchMessage(Handler.java:98) 
                   at android.os.Looper.loop(Looper.java:136) 
                   at android.os.HandlerThread.run(HandlerThread.java:61) 

Reply:
`Assertions.checkNotNull` throws an NPE which is what I meant but it looks like release builds have assertions disabled (assuming) so what you have is probably what we want. Though the `addVideoListener` should still happen before `client.open()` to avoid timing issues

Reply:
The rtsp negotiation is done in background (in Send and Reader threads) so there aren't timing issues.
The client.open() call returns immediately.

Reply:
Ah, 👍

Reply:
You're welcome. Thanks to you guys for adding this support!